horovod/torch/__init__.py

```diff
@@ -202,13 +202,182 @@ def zero_grad(self):
+        #  start = current.copy()
+        #  step() -> computes 'current - \alpha.f(g)' where f is
+        #            optimizer logic and g is the gradient
+        #  delta = current-start
+        #  allreduce_(delta)
+        #  start += delta
+        #  current = start
+        # In order to suppport this logic using function hook to improve performance,
+        # we do:
+        # delta = (start - \alpha.f(g)) - start
+        #       = -\alpha.f(g)
+        # set start to zero and step computes -\alpha.f(g)
+        # where f is the underlying optimizer logic
+
+        name = self._parameter_names.get(p)
+        start = self._starting_models[p]
+    
+        stashed_params = []
+        for group in self.param_groups:
+            stashed_params.append(group['params'])
+            # only want to step on p
+            if any([p is v for v in group['params']]):
+                group['params'] = [p]
+            else:
+                group['params'] = []
+
+        start.data.copy_(p)
+
+        super(self.__class__, self).step()
+
+        # compute delta = curr - start
+        p.data.sub_(start)
+        
+        # allreduce as before
+        tensor_compressed, ctx = self._compression.compress(p)
+        handle = allreduce_async_(tensor_compressed.data, name=name, op=Adasum)
+
+        # reset stashed parameters
+        for stashed, group in zip(stashed_params, self.param_groups):
+            group['params'] = stashed        
+
+        return handle, ctx
+
+    def _make_hook(self, p):
+        def hook(*ignore):
+            if p in self._handles and self._handles[p][0] is not None:
+                if self._allreduce_delay[p] <= 0:
+                    raise AssertionError(
+                        "Gradients were computed more than "
+                        "backward_passes_per_step times before call "
+                        "to step(). Increase backward_passes_per_step to "
+                        "accumulate gradients locally.")
+            assert not p.grad.requires_grad
+            assert self._allreduce_delay[p] > 0
+            handle, ctx = None, None
+            self._allreduce_delay[p] -= 1
+            if self._allreduce_delay[p] == 0:
+                handle, ctx = self._allreduce_grad_async(p)
+            self._handles[p] = (handle, ctx)
+        return hook
+
+    def synchronize(self):
+        pass
+
+    @contextmanager
+    def skip_synchronize(self):
+        raise AssertionError("Skipping synchronization is not supported when using Adasum optimizer.")
+
+    def step(self, closure=None):
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        missing_p = self._requires_update - set(self._handles.keys())
+        for p in missing_p:
+            handle, ctx = self._allreduce_grad_async(p)
+            self._handles[p] = (handle, ctx)
+
+        for p, (handle, ctx) in self._handles.items():
+            if not handle:
```

Comment:
I ran into an issue here when attempting to train through the `pytorch_imagenet_resnet50.py` script.  At the end of an epoch, the data may not divide evenly to allow for the exact number of batches per allreduce / step.  As a result, all the parameters had empty handles at the end of the epoch, and this `AssertionError` was raised.

In the original `DistributedOptimizer`, we force an allreduce any time `step()` is called to account for this possibility (see [here](https://github.com/horovod/horovod/blob/master/horovod/torch/__init__.py#L152)):

```
for p, value in self._handles.items():
            handle, ctx = value
            if handle is None:
                handle, ctx = self._allreduce_grad_async(p)
                self._handles[p] = (handle, ctx)
```

Can we not do the same here?

Reply:
Yes you are right. I didn't think of this case earlier.